### PR TITLE
#12179: Substitute std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>> with Program& 

### PR DESCRIPTION
**Pull request by rfurko-tt**
### Ticket
[Github ticket](https://github.com/tenstorrent/tt-metal/issues/12179)

### Problem description
There are complicated a few places which overuse `std::variant`, where it's not required and makes code hard to read and modify. To tackle this problem, we decided to simplify API to use `Program&` as a main way to propagate program. 

### What's changed
Change API and it's usage across the repository. 

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by patrickroberts** [comment]
If you changed the parameter type from `Program *` to `Program &`, there would be a lot less necessary changes, and the few shared_ptr unwrapping call sites would just be `*simple_program.get()` (notice the `*`). The code did not check for `nullptr` before so the refactored code without `std::visit` should be able to assume no empty shared_ptr as well, thus accepting by reference.

**Comment by rfurko-tt** [comment]
Discussed offline with @patrickroberts. Will change to Program&, but keep pointer in CommandInterface.

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 1]
```diff
@@ -2554,21 +2554,12 @@ void HWCommandQueue::terminate() {
...
-                    [&b](auto&& p) {
-                        using program_type = std::decay_t<decltype(p)>;
-                        if constexpr (std::is_same_v<program_type, std::reference_wrapper<Program>>) {
-                            p.get().add_buffer(b);
-                        } else {
-                            p->add_buffer(b);
-                        }
-                    },
-                    program);
+                program.add_buffer(b);
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:2562]: ❤️

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by patrickroberts — APPROVED** [review]

**Review by yan-zaretskiy — APPROVED** [review]

File: tt_metal/host_api.hpp [hunk 2]
```diff
@@ -264,8 +264,7 @@ void DeallocateBuffer(Buffer &buffer);
 *  | buffer   | The buffer that will be owned by the program | std::shared_ptr<Buffer> buffer |             | Yes      |
 *  | program  | The program getting ownership of the buffer  | std::shared_ptr<Buffer> buffer |             | Yes      |
 */
-void AssignGlobalBufferToProgram(
-    std::shared_ptr<Buffer> buffer, std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>> program);
+void AssignGlobalBufferToProgram(std::shared_ptr<Buffer> buffer, Program& program);
```
[the remark this answers is not in the document]
Comment (rfurko-tt) [reply tt_metal/host_api.hpp:267]: Thanks for noticing. Adjusted!

**Review by tt-asaigal — COMMENTED** [review]
Hey @rfurko-tt thanks for the cleanup :) Changes look good. @ayerofieiev-tt  We should  double check with the Moreh team about their usage of the `EnqueueProgram` API, they may be passing programs in as shared pointers and this may cause issues for them. Otherwise, I have no issues here.

**Comment by ayerofieiev-tt** [comment]
@tt-asaigal thank you! I will inform them. I propose to proceed and if they report any issues, we can revert/revisit/fix.

**Review by tt-asaigal — APPROVED** [review]

**Pull request merged**